bindings/rust/s2n-tls/src/renegotiate.rs

```diff
@@ -0,0 +1,677 @@
+        }
+    }
+}
+
+impl config::Builder {
+    /// Sets a method to be called when the client receives a request to renegotiate.
+    ///
+    /// See s2n_config_set_renegotiate_request_cb in [the C API documentation](https://github.com/aws/s2n-tls/blob/main/api/unstable/renegotiate.h).
+    pub fn set_renegotiate_callback<T: 'static + RenegotiateCallback>(
+        &mut self,
+        handler: T,
+    ) -> Result<&mut Self, Error> {
+        unsafe extern "C" fn renegotiate_cb(
+            conn_ptr: *mut s2n_connection,
+            _context: *mut libc::c_void,
+            response: *mut s2n_renegotiate_response::Type,
+        ) -> libc::c_int {
+            with_context(conn_ptr, |conn, context| {
+                let callback = context.renegotiate.as_mut();
+                if let Some(callback) = callback {
+                    if let Some(result) = callback.on_renegotiate_request(conn) {
+                        *response = result.into();
+                        return CallbackResult::Success.into();
+                    }
+                }
+                return CallbackResult::Failure.into();
+            })
+        }
+
+        let handler = Box::new(handler);
+        let context = self.config.context_mut();
+        context.renegotiate = Some(handler);
+        unsafe {
+            s2n_config_set_renegotiate_request_cb(
+                self.as_mut_ptr(),
+                Some(renegotiate_cb),
+                std::ptr::null_mut(),
+            )
+            .into_result()?;
+        }
+        Ok(self)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{
+        callbacks::{
+            ConnectionFuture, ConnectionFutureResult, PrivateKeyCallback, PrivateKeyOperation,
+        },
+        config::ConnectionInitializer,
+        testing::{CertKeyPair, InsecureAcceptAllCertificatesHandler, TestPair, TestPairIO},
+    };
+    use foreign_types::ForeignTypeRef;
+    use futures_test::task::new_count_waker;
+    use openssl::ssl::{
+        ErrorCode, NameType, Ssl, SslContext, SslFiletype, SslMethod, SslStream, SslVerifyMode,
+        SslVersion,
+    };
+    use std::{
+        error::Error,
+        io::{Read, Write},
+        pin::Pin,
+        task::Poll::{Pending, Ready},
+    };
+
+    // Currently renegotiation is not available from the openssl-sys bindings
+    extern "C" {
+        fn SSL_renegotiate(s: *mut openssl_sys::SSL) -> libc::size_t;
+        fn SSL_renegotiate_pending(s: *mut openssl_sys::SSL) -> libc::size_t;
+    }
+
+    // std::task::ready is unstable
+    fn unwrap_poll<T>(
+        poll: Poll<Result<T, crate::error::Error>>,
+    ) -> Result<(), crate::error::Error> {
+        if let Ready(value) = poll {
+            return value.map(|_| ());
+        }
+        panic!("Poll not Ready");
+    }
+
+    #[derive(Debug)]
+    struct ServerTestStream(TestPairIO);
+
+    // For server testing purposes, we read from the client output stream
+    impl Read for ServerTestStream {
```

Comment:
No, it's server specific. You have to implement the reverse stream for a client, because you read and write from the opposite buffers. So we'd need a separate ClientTestStream if I wanted to create an openssl client.

Reply:
Or I suppose you could have a generic in/out, but then you really can't use the same struct for the TestPair because the TestPair needs them specific for the client/server.

Reply:
I don't think it's specifically useful in this context, but if we find ourselves doing stuff like this relatively frequently it would probably be worth investigating what it would look like to unify the individual harnesses and the [benchmark harness](https://github.com/jmayclin/s2n-tls/blob/8ba2931d065dd31439f7927984cc197c4410b264/bindings/rust/bench/src/harness.rs#L268-L290).

So we'd just have a single harness that lives in `bindings/s2n-tls`, and then the benchmarks could take a dependency on that.